File: services/learncard-wallet-adapter/src/learncard_wallet_adapter/resultmap.py
"""Normalize LearnCard delivery outcomes into the adapter response (design §2)."""

from __future__ import annotations

from learncard_wallet_adapter.schemas import (
    DeliverRequest,
    DeliverResponse,
    DeliverResult,
    ErrorInfo,
)


def to_success(req: DeliverRequest, external_reference_id: str) -> DeliverResponse:
    return DeliverResponse(
        status="succeeded",
        workflow_id=req.workflow_id,
        execution_id=req.execution_id,
        step_id=req.step_id,
        correlation_id=req.correlation_id,
        external_reference_id=external_reference_id,
        result=DeliverResult(delivery_state="accepted"),
    )


def to_error(req: DeliverRequest, message: str) -> DeliverResponse:
    return DeliverResponse(
        status="failed",
        workflow_id=req.workflow_id,
        execution_id=req.execution_id,
        step_id=req.step_id,
        correlation_id=req.correlation_id,
        error=ErrorInfo(message=message),
    )

File: services/learncard-issuer-adapter/src/resultmap.ts
import { type IssuanceResult } from "./learncard";
import { type IssueResponse } from "./schemas";

// Normalize LearnCard SDK results + errors into the adapter response shape
// (design §3 `resultmap/`, §4 step 8). Keeps router-facing output stable as the
// SDK's own shapes change.

export function toSuccess(result: IssuanceResult): IssueResponse {
  return {
    status: "succeeded",
    external_reference_id: result.externalReferenceId,
    result: { issued_credential: result.issuedCredential },
    error: null,
  };
}

export function toError(err: unknown): IssueResponse {
  const message = err instanceof Error ? err.message : String(err);
  const code =
    typeof err === "object" && err !== null && "code" in err
      ? String((err as { code: unknown }).code)
      : undefined;
  return {
    status: "failed",
    external_reference_id: null,
    result: null,
    error: code ? { message, code } : { message },
  };
}
